### ISSUE TITLE
Race condition when using --hpx:connect?

### ISSUE THREAD

**Issue by biddisco** [STEllAR-GROUP/hpx#1684]
When adding nodes to a running job, my console node is querying the idle counters on each node to see if they need more work - when one connects, sometimes all is fine, other times I get an exception of type
```
{what}: no create function for performance counter found: \
     /threads{locality#2/total}/idle-rate \
    (counter type /threads/idle-rate is not defined, known counter types:
```
I presume that the performance counter is being queried before it has registered itself on the remote node (and with the console).

Is this a race - or should I wait somehow until the new node has finished initializing itself before attempting to do anything on it?

**Comment by hkaiser** [comment]
I think we should encapsulate the functionality of attaching a node to a running application from inside the application itself as standard functionality to HPX. This would allow to reliably avoid this kind of race conditions.

    future<id_type> add_locality(...);

The returned future would become ready once the newly attached locality is up and running.

**Comment by biddisco** [comment]
In your example above - is add_locality(...) called by the console node to bring a new node into the fold, or is add_locality(..., console, ...) called by a new worker that is created elsewhere.
In my own version, add_locality(...) is called by the console node and spawns processes - but the functionality to do this uses a number of process create/detach calls which are beyond the scope of hpx.
Therefore we can only really consider the second option. 
I'll look into this.

**Comment by hkaiser** [comment]
@biddisco I was hoping we could get the first version. But I understand that this is highly dependent on the job scheduling environment available.

> In my own version, add_locality(...) is called by the console node and spawns processes - but the functionality to do this uses a number of process create/detach calls which are beyond the scope of hpx.

That's why we have components. They allow to encapsulate things which are either proprietary, system dependent, or out of scope for the main libraries. If a component is installed its functionality is available, otherwise not.

**Comment by biddisco** [comment]
ok, understood, but then we have the problem that I'm using libraries that are not boost licensed and bringing them into the tree creates licensing issues we'd like to avoid.

If components are ok using the kind of approach I'm using with GitExternal for the verbs stuff, then I can look into it.

**Comment by hkaiser** [comment]
@biddisco Part of this will be solved by merging #2028.

**Comment by hkaiser** [comment]
> When adding nodes to a running job, my console node is querying the idle counters on each node to see if they need more work - when one connects, sometimes all is fine, other times I get an exception of type
>
>     {what}: no create function for performance counter found: \
>         /threads{locality#2/total}/idle-rate \
>        (counter type /threads/idle-rate is not defined, known counter types:
>
> I presume that the performance counter is being queried before it has registered itself on the remote node (and with the console).

The race condition itself has been solved by #2028. As part of that fix we added a means to force the launched application to optionally wait on a latch which allows to synchronize with the launching executable. Please see the tests [launch_process](https://github.com/STEllAR-GROUP/hpx/blob/master/tests/unit/component/launch_process.cpp) and [launched_process](https://github.com/STEllAR-GROUP/hpx/blob/master/tests/unit/component/launched_process.cpp) as examples on how to achieve that.

I'll go ahead and close this ticket for now. Please reopen if necessary.
